pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -178,6 +182,25 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
         # state dict.
         return super().lightning_module_state_dict()
 
+    def serialized_restore_model_state(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
+        checkpoint = {}
+        rank_zero_info(
+            f"FullyShardedDataParallel has {self.num_processes} processes. Serializing model
+            state restore to avoid CPU OOMs."
+        )
+        for current_worker in range(self.num_processes):
```

Comment:
Would this break if the number of gpus on checkpointing / saving isn't the same. Should we save the world_size + rank in each checkpoint and re-use that information on reload ?

Reply:
thanks for raising this! thinking aloud — in the same training session we wouldn't run into this problem, but even in the case where a model checkpoint from a different training session is being finetuned in a new training session with a different machine, do we envision that world size would be different (for our use cases at least we have config files that keep constant variables like Trainer(gpus) for the same model training)?

Reply:
I think that is definitely something we need to support. For example I usually use the output from `torch.cuda.device_count()` as number of gpus.

Reply:
@tchaton this kind of training "metadata" should get saved with the checkpoint. For example, we will also want to know this for fault-tolerance to fail if the trainer configuration has changed between runs and the user is trying to restore mid-batch.

Reply:
FWIW i ran a test taking a checkpoint from a previous training session and starting a fresh one: setting trainer.resume_from_checkpoint to the old checkpoint and trainer.gpus=4 instead of 8 (which was num gpus on the original training session) and loading the checkpoint on 4 GPUs didn't break. 

i can add this feature in another pull request! would a good place to add the metadata be `on_save_checkpoint` in `model_checkpoint.py`? eg here: https://github.com/PyTorchLightning/pytorch-lightning/blob/c7f8c8c3c82b4f249125885490b2392bf9d3d08b/pytorch_lightning/callbacks/model_checkpoint.py#L341